### ISSUE TITLE
Reconsider default alternator_streams_time_window_s = 10 seconds

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#6929]
Alternator's Streams implementation uses a parameter, `alternator_streams_time_window_s`, which defaults to 10 seconds, . What this 10 second interval does is that CDC log entries newer than 10 seconds old will not be visible. If I remember correctly, we claimed that we need this delayed reading to allow delayed CDC log writes to be reordered to the right order - and this is the behavior we recommend for all CDC clients - although this recommendation is missing in 
https://docs.scylladb.com/using-scylla/cdc/cdc-querying-streams/ (CC @haaawk, FYI). 

This 10-second latency is annoying for tests - it makes every test 10 second long (in Alternator's tests, this is normally the time that we run 100 tests ;-)). But if it was just tests, I would just run the tests with a zero `alternator_streams_time_window_s` and that's it (I will probably do that, no matter what we decide in this issue). What worries me more is whether users will accept this: 

Apparently, users care about CDC latency, and in https://stackoverflow.com/questions/41965786/how-real-time-dynamodb-stream-is a user suggested that DynamoDB Streams have a latency of less than half a second (I'll try later to write a test to confirm that estimate). I wanted to raise here the question of whether a 10 second delay in Alternator Streams is acceptable, reasonable, or even necessary. Under what circumstances could CDC entries be reordered for 10 seconds? In such exceptional circumstances - couldn't they be delayed also for 1000 seconds? I think that if the answer to my last question is "yes" it means the 10 second number should not have been chosen, and we should choose a much lower "happy path" number. My reasoning is that if the "10 seconds" number isn't guaranteed to be high enough for all exceptional cases, then what does it really guarantee?

CC @elcallio @haaawk @kbr-

**Comment by nyh** [comment]
The 10-second delay has another weird consequence that I thought is worth mentioning:

In the Streams API `GetShardIterator`, the user can ask for a shard _iterator_ with the time `LATEST`. This option means (see https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_streams_GetShardIterator.html): "Start reading just after the most recent stream record in the shard". With our 10 second delay, "most recent stream record" actually means whatever was written 10 seconds ago. This means that a write that was done, say, 5 seconds before asking for the `LATEST`iterator, will be returned by that iterator! This is inconvenient for a user who hopes that using `LATEST` iterator means he would only see new writes, not past writes. Arguably, this user's hope is impossible (e.g., a write that happened just one millisecond ago definitely might appear a millisecond late and get into the iterator) but from one millisecond to 10 seconds, there is a long way.

**Comment by elcallio** [comment]
The problem as always are client timestamps. Otoh, since these don't exist in alternator (right?), the only _real_ issue is mismatched clocks in cluster. So there is a point for maybe allowing a smaller window for alternator. 

For "normal" CDC we need the confidence window to deal with client timestamps reordering the stream events. Otherwise we could miss data, since we don't have monotonic counters (cross-node!), but rely on the timestamps as markers and positions.

**Comment by kbr-** [comment]
> Under what circumstances could CDC entries be reordered for 10 seconds? In such exceptional circumstances - couldn't they be delayed also for 1000 seconds? I think that if the answer to my last question is "yes" it means the 10 second number should not have been chosen, and we should choose a much lower "happy path" number. My reasoning is that if the "10 seconds" number isn't guaranteed to be high enough for all exceptional cases, then what does it really guarantee?

In theory, clock drifts are unbounded, and there are no guarantees.

In practice, it's all statistics. AFAIK Scylla clusters use NTP for clock synchronization, and the longer interval we take, the higher probability that clock drifts will fit in this interval.

In this paper: https://www.rfc-editor.org/rfc/rfc1128.pdf
the results of a survey are presented with some graphs attached showing the measured probabilities.
Look at section 2.1 and figure 2:
![image](https://user-images.githubusercontent.com/6516501/88646258-42701d80-d0c5-11ea-8244-241133926f9a.png)
(the bottom curve is for NTP):
- `P(offset > 1sec) < 0.1`
- `P(offset > 10sec) < 0.01`
- `P(offset > 100sec)` is roughly the same as  `P(offset > 10sec)`

From the section:
>  In the case of NTP, for example, which is designed expressly for time synchronization, eight hosts showed errors above ten seconds, a value considered barely credible for a host synchronized by NTP in the Internet. It is very likely that some or all of these hosts, representing about one percent of the total NTP population, were using an old NTP implementation with known bugs.

**Comment by kbr-** [comment]
> This 10-second latency is annoying for tests - it makes every test 10 second long 

Hah. Have you ever tried running a test from SCT? :D

**Comment by nyh** [comment]
@kbr- @elcallio  a few notes:

First I think it is not true that the only reason why the `streams_time_window` is needed is clock differences between the nodes. Let me explain:

The other potential cause of reordering is the latency of the network and replication: Imagine for a moment perfectly synchronized clocks. The client sends a request at time t1 to coordinator node1, and at time t2 to node2. Lets say that t1<t2, and the client happens to read the CDC log from a quorum of node2 and node3. The write t1 will arrive at node2 and node3 after some delay of typically multiple mililseconds - a delay that has nothing to do with clock drift. For this reason it is possible that the client will read in the CDC log the write at t2, but will not yet see the write at the earlier time of t1! It is this reordering that streams_time_window was meant to solve, I believe.

However, that being said, there is another thing I want to say:  Even if the reordering described in the above example happens, the resulting problem is relatively minor problem in Alternator. The DynamoDB API (deliberately) does not support deltas in the CDC, just full item images. The result of the reordering described above is that the client will read the event at t2, and then will not see the event at t1 because it is earlier (and the next request will ask for time >t2). But so what? Since t2 is later, the NEW_IMAGE at t2 is the latest one anyway. KEYS_ONLY is fine too. The only problem is the OLD_IMAGE, which misses the even older image from before t1. If (?) we care about this OLD_IMAGE problem - or want to be sure that even for KEYS_ONLY two writes lead two two events (who cares?) - then we should care about the reordering problem in Alternator too.
   
Anyway, if it is the replication delays that we are worried about - not the clock drift - then I have some things to say about it:

1. It is "usually" in the order of magnitude of milli-seconds.
2. In the worst case, it can be unbounded. E.g., A (failed) write may reach just one replica and be replicated by a repair hours later.
3. For *successful* writes, we have a bound on this replication delay (for a quorum read!!), I think it is 2 seconds.

I understand that the probability of a problem with a threshold of 10 seconds is higher than that with a threshold of 1 second (the https://en.wikipedia.org/wiki/Cumulative_distribution_function is non-decreasing ;-)) but that's just a tautology and doesn't help us decide between 10 seconds or 1 seconds or 100 seconds.

Finally, @kbr- , one note about NTP: I don't know what this graph you shown us is about, I didn't read the source (at least not in this millennium), but it seems completely **wrong** or just irrelevant to modern computers. The bottom line of the graph suggests that a whopping 1% of the nodes which are using NTP (?!) are not only off by a whole second - almost all of these 1% are off by what appears to be exactly the same amount around 2 minutes. This makes absolutely no sense.... Nodes which use NTP keep synchronized to the order of milliseconds, and even if they loose their network connectivity they still have an internal clock which is pretty accurate - and it would take months, if not years, of dis-connectivity for the clock to drift by as much as a few minutes. You will definitely never see 1% of your nodes drifting by two minutes, as this graph suggests. I think these 1% of the nodes are just blatantly misconfigured and aren't actually using NTP as they think.

P.S. Did you guys ever document the reordering problem - its causes, implications, and what to do about it (the time_window thing) - in the CDC user documentation?

**Comment by kbr-** [comment]
@nyh 

> The other potential cause of reordering is the latency of the network and replication

True, and as you explain, it's even worse than clock drift (for failed writes at least).

> but that's just a tautology and doesn't help us decide between 10 seconds or 1 seconds or 100 seconds.

The CDF's **rate of increasing** helps us decide.
I wanted to point out that between 1s and 10s there is an **order of magnitude** difference in probability (0.1 vs 0.01), and between 10s and 100s the difference is insignificant.

> Since t2 is later, the NEW_IMAGE at t2 is the latest one anyway.

The missing NEW_IMAGE from t1 could be related to a different key.
But if the keys are different, the user will do fine by performing a requery of this window, I guess.

> I don't know what this graph you shown us is about, I didn't read the source (at least not in this millennium), but it seems completely wrong or just irrelevant to modern computers. The bottom line of the graph suggests that a whopping 1% of the nodes which are using NTP (?!) are not only off by a whole second - almost all of these 1% are off by what appears to be exactly the same amount around 2 minutes. This makes absolutely no sense.... Nodes which use NTP keep synchronized to the order of milliseconds, and even if they loose their network connectivity they still have an internal clock which is pretty accurate - and it would take months, if not years, of dis-connectivity for the clock to drift by as much as a few minutes. You will definitely never see 1% of your nodes drifting by two minutes, as this graph suggests. I think these 1% of the nodes are just blatantly misconfigured and aren't actually using NTP as they think.

Yes, and the authors agree with you. From the paper:
> In the case of NTP, for example, which is designed expressly for time synchronization, eight hosts showed errors above ten seconds, a value considered barely credible for a host synchronized by NTP in the Internet. It is very likely that some or all of these hosts, representing about one percent of the total NTP population, were using an old NTP implementation with known bugs.

**Comment by nyh** [comment]
> I wanted to point out that between 1s and 10s there is an **order of magnitude** difference in probability (0.1 vs 0.01), and between 10s and 100s the difference is insignificant.

That is true, but we can't base this decision on some graph from 30 years ago based on just a few hundred measurements that were probably incorrect even at the time. And even if that graph was accurate it doesn't include the issue of replication latency.

> > Since t2 is later, the NEW_IMAGE at t2 is the latest one anyway.
> 
> The missing NEW_IMAGE from t1 could be related to a different key.

This is a very good point. So yes, missing that write from t1 could be missing any write of any key, completely unrelated to what happened in t2, and is indeed very wrong even for Alternator. So yes, we need to guarantee that it doesn't happen. The document we should have about the need for this "time window" thing should clearly explain this - that what can happen if you don't employ some sort of "time_window" solution is that you can lose any write.

> But if the keys are different, the user will do fine by performing a requery of this window, I guess.

I guess you mean here the alternative solution mentioned by @haaawk on today's call - that the next query asks to start at the last known timestamp minus some time window (10 seconds or whatever). Yes, this is also a workable solution, but more work for the client implementation - and needs to hold in memory the CDC-log keys of all the events seen in the last time window so the client doesn't process the same event again.

**Comment by kbr-** [comment]
> but we can't base this decision on some graph from 30 years ago based on just a few hundred measurements that were probably incorrect even at the time.

True, but it's even worse to base our decision on guessing.
I think we need to make an educated decision based on empirical data. I agree that the data I linked is probably outdated -- it's just what I found after a minute of googling -- but I'd assume there exists more recent research in this domain.

**Comment by ScyllaPiotr** [comment]
One the one hand, we could leave the value as it is, but on the other, this is risky - like @nyh said, we do not know if the value is accurately selected. I believe we do not want to learn this only after a client starts using streams in alternator. I suppose until they are GA, no one uses them for real, but I asked out to be sure.

From a more technical standpoint, I am yet to understand if the reordering of CDC log is unbounded or not. One possible explanation is that it is bounded by the topology reordering timeout of 2 * _ring delay_ (default 30s) + _generation leeway_ (default 5s), so a total of 65s. But is it actually a maximum? More importantly, can a customer using Alternator even cause/experience any changes to topology?

Other that that, @nyh wrote the following about replication delays:
> 1. It is "usually" in the order of magnitude of milli-seconds.
> 2. In the worst case, it can be unbounded. E.g., A (failed) write may reach just one replica and be replicated by a repair hours later.

In Alternator we write with CL=LOCAL_QUORUM. With this in mind, a failed write reaching only one replica and persisting there should be very rare.

> 3. For successful writes, we have a bound on this replication delay (for a quorum read!!), I think it is 2 seconds.

`storage_proxy::do_query()` inside `service/storage_proxy.cc` gets default Alternator timeout value of 10s inside `query_options`. It seems this value is used with all the requests further down the call stack.
If 10s read timeout is used, perhaps we should use like _10s+2s_ for the `alternator_streams_time_window_s`?

**Comment by nyh** [comment]
@ScyllaPiotr I don't remember where I got the number 2 seconds from (I wrote it 5 years ago). Actually as you said Alternator uses  default_timeout_in_ms for all operations - both reads and writes, and this defaults to 10 seconds.

This means that **in theory**, we could have an Alternator write operation that takes 9.9 seconds because the write was immediate in replica A and then delayed mysteriously by 9.9 seconds before written to replica B replica C. This means that the CDC entry can also be delayed by 9.9 seconds. If a second write was done a millisecond later and somehow happened _immediately_ on the three replicas, we can have a read from the CDC log see the second write 9.9 seconds before the first write.

So it sounds to me like a 10s alternator_streams_time_window_s is actually correct in that sense.

I wonder if we shouldn't reduce both default default_timeout_in_ms and alternator_streams_time_window_s - 10s sounds a bit too much for both.

Alternatively we could do something like described in https://github.com/scylladb/scylladb/issues/6929#issuecomment-665588660 - have zero window, but make read requests go back in time and de-duplicate already returned events. But this solution is not trivial to implement, and could also benefit from starting with a lower number than 10 seconds.

**Comment by ScyllaPiotr** [comment]
Using exactly 10s (or the same timeout value in general) leaves no room for Scylla to actually acknowledge that another 10s timeout (`default_timeout_in_ms`) actually passed or not in the corner case of 9.(9) sec. That's why I proposed 10s+2s. But maybe I'm being overly cautious here.

**Comment by nyh** [comment]
@swasik noticed that the `alternator_streams_time_window_s` parameter is not even documented. docs/alternator/compatibility.md says:

> While in DynamoDB data usually appears in the stream less than a second after it was written, in Alternator Streams there is currently a 10 second delay by default.

But doesn't mention how this 10 can be configured. If we do mention how it can be configured with `alternator_streams_time_window_s`, we also need to explain how it should be properly chosen - explaining this won't be easy (see discussions above).

**Comment by kadishmal** [comment]
Hello, team :wave:!

We're evaluating ScyllaDB as an alternative to DynamoDB. We are a FinTech company heavily relying on ACID transactions support by DynamoDB as well as near-real-time Stream of changes.

Since this issue specifically discusses the Streams, let me provide some input for your consideration.

As a Payments Service Provider (PSP) we receive a payment request against an upstream Service Provider and store it in the table with the `PENDING` status. As soon as the request is PUT, almost instantly we receive the event from the Stream and relay it to the remote Service Provider, further moving the payment to `IN_PROGRESS` state. Then this change triggers another processor which finally changes the payment state to `SUCCEEDED`/`FAILED` state. Thus, the end goal is to complete processing the user request as soon as possible. With DynamoDB we end up with less than a second delay in between the processes, overall completing the E2E payment in about 5-7 seconds including the original request as well as the upstream communications.

Considering the nature of PSP, frankly speaking, it will definitely become unacceptable for users to wait for 20+ seconds for a payment to complete. If `alternator_streams_time_window` parameter is configurable, we would set it to 0.5 seconds at most. Anything above 1 second may pose a threat to our business in terms of disadvantage when compared to competitors. Here I have couple of questions:

1. Is this option configurable?
2. Is there a minimum value we cannot set the value below than that?
3. Will setting this value to 0.5s provide the behaviour we expect as described in the DynamoDB case above.

Though, the multi-item ACID transactions support is the main deal-breaker which completely blocks us from migrating to ScyllaDB, even when the support is finally added, we will still have a difficult time migrating to ScyllaDB due to this CDC delay.

It is a fact that I may not know some implementation and administrative details of ScyllaDB, I would really appreciate if you could, perhaps, provide a solution to the above two problems we face with migrating to ScyllaDB.

Thank you.

**Comment by nyh** [comment]
Hi @kadishmal, thanks for sharing your use case.

> If `alternator_streams_time_window` parameter is configurable, we would set it to 0.5 seconds at most. Anything above 1 second may pose a threat to our business in terms of disadvantage when compared to competitors. Here I have couple of questions:
> 
>  1. Is this option configurable?
>  2. Is there a minimum value we cannot set the value below than that?
>  3. Will setting this value to 0.5s provide the behaviour we expect as described in the DynamoDB case above.

`alternator_streams_time_window_s` is indeed configurable. Unfortunately, you can't actually set it to 0.5 seconds because... it's type is integer and it's in seconds instead of milliseconds :-( This strange oversight can be easily fixed because internally we work with milliseconds, not seconds.

So, would setting it to 1 seconds work? Well, "sort of":
In the happy case, it will work just fine. Normally request latency, even P99 latency, is just a few milliseconds. A full second or half a second is "eons", and if you look at what happened half a second ago, you'll easy see everything.
But... This is not guaranteed. Above I gave an example problematic scenario: Consider that you had a 2-second network hiccup. One write before that hiccup managed to one to one replica, but not yet to the two others. This write hasn't yet given up (because `alternator_timeout_in_ms` defaults to 10 seconds) and will retry, but before it finally succeeds a second write quickly does all three updates. The end result is that the events enter the streams table in the wrong order. Which isn't a problem if you only read it later - but if the application reads from it too quickly and happen to read the second event first, when the application wants to continue reading, it will skip past the second event - and miss the first event.
Lowering also `alternator_timeout_in_ms` will solve this scenario, but as I mentioned above, I'm not completely sure solves all scenarios (I'm not one of the ScyllaDB CDC developers, so I am not really the expert in this area).

But unfortunately, there's a second problem beyond `alternator_streams_time_window_s` - a second problem that makes it **inefficient** to get very low event-latencies from Alternator streams. The problem is the often huge number of "shards" used by Alternator streams (https://github.com/scylladb/scylladb/issues/13080). Imagine that you application needs poll 10,000 stream "shards" (and not just 3 or 10) to get the next events. Now imagine needing to do this every 0.5 seconds if you want to be sure events don't get delayed by more than 0.5 seconds. It's not impossible, but it's definitely a problem. 

We will definitely be working on all three issues (this one, https://github.com/scylladb/scylladb/issues/13080, and https://github.com/scylladb/scylladb/issues/5064) eventually, but unfortunately can't really say when.
